### PR TITLE
WESTPA 2.0 Tests, Consolidated

### PR DESCRIPTION
**Pull request by jeremyleung521**
**Issue Number.** Is this pull request related to any outstanding issues? If so, list the issue number.  
#120 #124 #126 #135 #123 #127 #134 #148 

**Describe the changes made.** A clear and concise description of what the problem is and what you did to fix it. E.g. [...] was happening and I've changed [...] to fix it.  
A branch consolidating all the tests made. This is to eliminate duplicate reference files/fixtures/variable names and unify test codes.

**Goals and Outstanding Issues.** A clear and concise list of goals (to be) accomplished.  
- [x] plothist (@hep32, #126)
- [x] w_succ (@darianyang, #120)
- [x] w_pdist (@jeremyleung521, #124)
- [x] Update w_trace
- [x] Update w_ipa
- [x] Incorporate new w_assign (@atbogetti, #135)
- [x] Incorporate w_direct 
- [x] Incorporate Data Manager Tests (@poharrison, #127)
- [x] Incorporate w_init, w_run, w_truncate (@jdrusso, #123)  
- [x] fix w_states
- [x] CI miniconda updated to conda-incubator/setup-miniconda@v2 to accomodate changes in github action
- [x] Force CI numpy version to <=1.18.5 until #129 is fixed
- [x] Fixed odd case in states.py where auxref and state.label are bytestrings and causes states_to_file to fail -- Check below for context (Thanks @jdrusso) 
- [x] Sim manager tests (@dnlebard, #134, #148)

Thank you, everyone, for your tests. 

Current Issues:
- w_direct and w_ipa do not use mock.patch (Involves too many functions down the line)

**Major files changed.**  
- [x] all the tests in /tests/test_tools/
- [x] /tests/refs folder
- [x] /src/westpa/cli/core/w_init.py 
- [x] /tests/test_sim_manager.py
- [x] /.github/workflows/test.yaml
- [x] /devtools/conda-envs/test_env.yaml
- [x] /src/westpa/core/states.py
- [x] /tests/test_sim_manager.py
- [x] /tests/test_we_driver.py

**Status.**
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

**Additional context.** Add any other context or screenshots about the pull request here.  
- About the problem in states.py, it seems somewhere between _init_ and states_to_file, auxref turns into "b''". states.label is also initialized as a bytestring. If statements are added to states.py to ensure these edge cases don't occur. This also fixes the w_states test errors that happen exclusively on GH Actions.
- This PR includes a patch from @synapticarbors (https://github.com/synapticarbors/westpa/commit/b2e3f2020a8991013bcc05f4aee8821a45779dca.patch), and fixes issues with race conditions on tests originating from simulation manager tests.

### CODE REVIEW

**Comment by jeremyleung521** [comment]
Test_w_state still failing in github despite no changes from a previously successful CI run (and not reproducible locally). Closing this as I investigate why.

File: tests/test_sim_manager.py [hunk 1]
```diff
@@ -20,38 +26,79 @@ def setup(self):
...
+        westpa.rc._data_manager = None
         del os.environ['WEST_SIM_ROOT']
-
-    def test_sim_manager(self):
-        assert self.sim_manager.n_propagated == 0
-        assert len(self.sim_manager._callback_table) == 0
+        # del self.sim_manager
+        # shutil.rmtree(self.test_dir)
+        # data = westpa.rc.new_data_manager()
+        # data.we_h5filename = None
```
Comment (dnlebard) [inline tests/test_sim_manager.py:69]: Extremely minor nitpick: we should delete the commented out stuff.

File: tests/test_tools/conftest.py [hunk 2]
```diff
@@ -18,37 +20,112 @@
...
+    os.chdir(test_dir)
+
+    copy_ref(test_dir)
+
+    copyfile(os.path.join(REFERENCE_PATH, 'west_3iter.h5'), H5_FILENAME)
+    copyfile(os.path.join(REFERENCE_PATH, 'west_init_ref.cfg'), CFG_FILENAME)
 
-    copyfile(refs_h5_file, H5_FILENAME)
-    copyfile(refs_cfg_file, CFG_FILENAME)
+    # copyfile(os.path.join(REFERENCE_PATH, 'west_init_ref.h5'), 'west_init_ref.h5')
```
Comment (dnlebard) [inline tests/test_tools/conftest.py:43]: Some leftover debugging to delete.

File: tests/test_tools/conftest.py [hunk 3]
```diff
@@ -18,37 +20,112 @@
...
-    os.remove(H5_FILENAME)
+    test_dir = str(tmpdir)
+    os.chdir(test_dir)
 
-    os.chdir(STARTING_PATH)
+    copy_ref(test_dir)
+
+    copyfile(os.path.join(REFERENCE_PATH, 'west_init_ref.cfg'), CFG_FILENAME)
+    copyfile(os.path.join(REFERENCE_PATH, 'west_init_ref.h5'), "west_init_ref.h5")
+    # copyfile(os.path.join(REFERENCE_PATH, H5_FILENAME), H5_FILENAME)
```
Comment (dnlebard) [inline tests/test_tools/conftest.py:68]: to be deleted.

File: tests/test_tools/test_w_trace.py [hunk 4]
```diff
@@ -59,6 +56,3 @@ def check_output(self, arg_idx):
             assert list(traj_group.keys()) == expected_groups, "H5 groups ({}) are not as expected ({})".format(
                 list(traj_group.keys()), expected_groups
             )
-
-        diff = H5Diff(self.ref_files[arg_idx], self.outfile)
-        diff.check()
```
Comment (dnlebard) [inline tests/test_tools/test_w_trace.py:64]: Do we not need to check against the index anymore? Just double checking the logical change was warranted.

**Review by dnlebard — COMMENTED** [review]
Overall this looks good, and I can't find any real issues. Just small nit-picky things like leftover debug comments.

File: tests/test_tools/test_w_trace.py (hunk 4 above)
Comment (jeremyleung521) [reply tests/test_tools/test_w_trace.py]: Seems like the test was never updated to the newer version with H5Diff. Thanks for noticing this! Now fixed.
